pkg/ccl/serverccl/statusccl/statusccl_test: TestTenantStatusAPI failed

pkg/ccl/serverccl/statusccl/statusccl_test.TestTenantStatusAPI [failed](https://tanzanite.cluster.engflow.com/invocations/default/3248e418-de91-40d2-871c-3359f9d01e35?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3NlcnZlcmNjbC9zdGF0dXNjY2w6c3RhdHVzY2NsX3Rlc3Q=) on master @ [4f31fbae529e459c00ef2490e193c0c246e8d0f9](https://github.com/cockroachdb/cockroach/commits/4f31fbae529e459c00ef2490e193c0c246e8d0f9):


```
=== RUN   TestTenantStatusAPI
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestTenantStatusAPI2715621298
    tenant_status_test.go:141: -- test log scope end --
CREATE UNIQUE INDEX test_pkey ON test_db1.public.test USING btree (k ASC)test logs left over in: outputs.zip/logTestTenantStatusAPI2715621298
--- FAIL: TestTenantStatusAPI (1641.81s)
=== RUN   TestTenantStatusAPI/index_usage_stats
    tenant_status_test.go:1062: 
        	Error Trace:	pkg/ccl/serverccl/statusccl/statusccl_test/pkg/ccl/serverccl/statusccl/tenant_status_test.go:1062
        	            				pkg/ccl/serverccl/statusccl/statusccl_test/pkg/ccl/serverccl/statusccl/tenant_status_test.go:110
        	Error:      	Not equal: 
        	            	expected: [][]string{[]string{"118", "1", "2", "true"}, []string{"118", "2", "1", "true"}}
        	            	actual  : [][]string{[]string{"118", "1", "2", "false"}, []string{"118", "2", "1", "false"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -5,3 +5,3 @@
        	            	   (string) (len=1) "2",
        	            	-  (string) (len=4) "true"
        	            	+  (string) (len=5) "false"
        	            	  },
        	            	@@ -11,3 +11,3 @@
        	            	   (string) (len=1) "1",
        	            	-  (string) (len=4) "true"
        	            	+  (string) (len=5) "false"
        	            	  }
        	Test:       	TestTenantStatusAPI/index_usage_stats
    --- FAIL: TestTenantStatusAPI/index_usage_stats (119.31s)
=== RUN   TestTenantStatusAPI/tenant_auth_statement
    tenant_status_test.go:1452: 
        	Error Trace:	pkg/ccl/serverccl/statusccl/statusccl_test/pkg/ccl/serverccl/statusccl/tenant_status_test.go:1452
        	            				pkg/ccl/serverccl/statusccl/statusccl_test/pkg/ccl/serverccl/statusccl/tenant_status_test.go:123
        	Error:      	Received unexpected error:
        	            	net/http: request canceled (Client.Timeout or context cancellation while reading body)
        	Test:       	TestTenantStatusAPI/tenant_auth_statement
    --- FAIL: TestTenantStatusAPI/tenant_auth_statement (45.18s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cluster-observability @cockroachdb/multi-tenant
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantStatusAPI.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-34988

pkg/ccl/serverccl/statusccl/statusccl_test.TestTenantStatusAPI [failed](https://tanzanite.cluster.engflow.com/invocations/default/9a7a0fad-4521-4f50-8729-bf6d8f8789cb?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3NlcnZlcmNjbC9zdGF0dXNjY2w6c3RhdHVzY2NsX3Rlc3Q=) on master @ [57bd0a92676b1c58ee7e29979ce3dceb5837abdb](https://github.com/cockroachdb/cockroach/commits/57bd0a92676b1c58ee7e29979ce3dceb5837abdb):


```
=== RUN   TestTenantStatusAPI
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestTenantStatusAPI3390548490
    tenant_status_test.go:141: -- test log scope end --
CREATE UNIQUE INDEX test_pkey ON test_db1.public.test USING btree (k ASC)test logs left over in: outputs.zip/logTestTenantStatusAPI3390548490
--- FAIL: TestTenantStatusAPI (971.46s)
=== RUN   TestTenantStatusAPI/reset_index_usage_stats/http
    sql_runner.go:87: error executing query="\nCREATE TABLE test (\n  k INT PRIMARY KEY,\n  a INT,\n  b INT,\n  INDEX(a)\n)\n" args=[]: pq: relation "defaultdb.public.test" already exists
        --- FAIL: TestTenantStatusAPI/reset_index_usage_stats/http (0.44s)
=== RUN   TestTenantStatusAPI/reset_index_usage_stats/sql-cli
    tenant_status_test.go:726: query '
        SELECT
          table_id,
          index_id,
          total_reads,
          extract_duration('second', now() - last_read) < 5
        FROM
          crdb_internal.index_usage_statistics
        WHERE
          table_id = 104': expected:
        104, 1, 2, true
        104, 2, 1, true
        
        got:
        104, 1, 2, false
        104, 2, 1, false
        
        --- FAIL: TestTenantStatusAPI/reset_index_usage_stats/sql-cli (14.79s)
=== RUN   TestTenantStatusAPI/reset_index_usage_stats
    --- FAIL: TestTenantStatusAPI/reset_index_usage_stats (15.25s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantStatusAPI.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/ccl/serverccl/statusccl/statusccl_test.TestTenantStatusAPI [failed](https://tanzanite.cluster.engflow.com/invocations/default/f4a84354-673d-45b5-87e7-b12209e8eb4b?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3NlcnZlcmNjbC9zdGF0dXNjY2w6c3RhdHVzY2NsX3Rlc3Q=) on master @ [c316d6a615fa02a05357a20bf03b8a197ab27810](https://github.com/cockroachdb/cockroach/commits/c316d6a615fa02a05357a20bf03b8a197ab27810):


```
  pkg/ccl/serverccl/statusccl/statusccl_test.TestTenantStatusAPI.func7()
      pkg/ccl/serverccl/statusccl/statusccl_test/pkg/ccl/serverccl/statusccl/tenant_status_test.go:102 +0x64
  testing.tRunner()
      GOROOT/src/testing/testing.go:1595 +0x261
  testing.(*T).Run.func1()
      GOROOT/src/testing/testing.go:1648 +0x44

Goroutine 272063 (running) created at:
  pkg/ccl/serverccl/statusccl/statusccl_test.testTenantStatusCancelQuery()
      pkg/ccl/serverccl/statusccl/statusccl_test/pkg/ccl/serverccl/statusccl/tenant_status_test.go:1184 +0x2f5
  pkg/ccl/serverccl/statusccl/statusccl_test.TestTenantStatusAPI.func7()
      pkg/ccl/serverccl/statusccl/statusccl_test/pkg/ccl/serverccl/statusccl/tenant_status_test.go:102 +0x64
  testing.tRunner()
      GOROOT/src/testing/testing.go:1595 +0x261
  testing.(*T).Run.func1()
      GOROOT/src/testing/testing.go:1648 +0x44

Goroutine 272062 (finished) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:1648 +0x845
  pkg/ccl/serverccl/statusccl/statusccl_test.TestTenantStatusAPI()
      pkg/ccl/serverccl/statusccl/statusccl_test/pkg/ccl/serverccl/statusccl/tenant_status_test.go:101 +0xafa
  github.com/cockroachdb/cockroach/pkg/ccl/serverccl.newTestTenant()
      github.com/cockroachdb/cockroach/pkg/ccl/serverccl/tenant_test_utils.go:98 +0xb6
  github.com/cockroachdb/cockroach/pkg/ccl/serverccl.newTenantClusterHelper()
      github.com/cockroachdb/cockroach/pkg/ccl/serverccl/tenant_test_utils.go:223 +0x17b
  github.com/cockroachdb/cockroach/pkg/server.(*testServer).StartTenant()
      github.com/cockroachdb/cockroach/pkg/server/testserver.go:1769 +0x2f7b
  github.com/cockroachdb/cockroach/pkg/testutils/serverutils.StartTenant()
      github.com/cockroachdb/cockroach/pkg/testutils/serverutils/test_server_shim.go:398 +0xa5
  github.com/cockroachdb/cockroach/pkg/ccl/serverccl.newTestTenant()
      github.com/cockroachdb/cockroach/pkg/ccl/serverccl/tenant_test_utils.go:98 +0xb6
  github.com/cockroachdb/cockroach/pkg/ccl/serverccl.newTenantClusterHelper()
      github.com/cockroachdb/cockroach/pkg/ccl/serverccl/tenant_test_utils.go:223 +0x17b
  github.com/cockroachdb/cockroach/pkg/ccl/serverccl.NewTestTenantHelper()
      github.com/cockroachdb/cockroach/pkg/ccl/serverccl/tenant_test_utils.go:156 +0x1fd
  github.com/cockroachdb/cockroach/pkg/ccl/serverccl.NewTestTenantHelper()
      github.com/cockroachdb/cockroach/pkg/ccl/serverccl/tenant_test_utils.go:146 +0xfc
  pkg/ccl/serverccl/statusccl/statusccl_test.TestTenantStatusAPI()
      pkg/ccl/serverccl/statusccl/statusccl_test/pkg/ccl/serverccl/statusccl/tenant_status_test.go:66 +0x364
  testing.tRunner()
      GOROOT/src/testing/testing.go:1595 +0x261
  testing.(*T).Run.func1()
      GOROOT/src/testing/testing.go:1648 +0x44
==================
=== RUN   TestTenantStatusAPI/tenant_cancel_session
    sql_runner.go:87: error executing query="SELECT 1" args=[]: sql: database is closed
=== RUN   TestTenantStatusAPI/reset_index_usage_stats
=== RUN   TestTenantStatusAPI/tenant_contention_event
    sql_runner.go:87: error executing query="CREATE DATABASE IF NOT EXISTS test;CREATE TABLE test.test (x INT PRIMARY KEY);" args=[]: sql: database is closed
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantStatusAPI.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/ccl/serverccl/statusccl/statusccl_test.TestTenantStatusAPI [failed](https://tanzanite.cluster.engflow.com/invocations/default/2c88008c-7b8d-494a-9a37-3f4860689843?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3NlcnZlcmNjbC9zdGF0dXNjY2w6c3RhdHVzY2NsX3Rlc3Q=) on master @ [c316d6a615fa02a05357a20bf03b8a197ab27810](https://github.com/cockroachdb/cockroach/commits/c316d6a615fa02a05357a20bf03b8a197ab27810):


```
=== RUN   TestTenantStatusAPI
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestTenantStatusAPI1827042902
    tenant_status_test.go:141: -- test log scope end --
CREATE UNIQUE INDEX test_pkey ON test_db1.public.test USING btree (k ASC)test logs left over in: outputs.zip/logTestTenantStatusAPI1827042902
--- FAIL: TestTenantStatusAPI (1283.72s)
=== RUN   TestTenantStatusAPI/index_usage_stats
    tenant_status_test.go:1062: 
        	Error Trace:	pkg/ccl/serverccl/statusccl/statusccl_test/pkg/ccl/serverccl/statusccl/tenant_status_test.go:1062
        	            				pkg/ccl/serverccl/statusccl/statusccl_test/pkg/ccl/serverccl/statusccl/tenant_status_test.go:110
        	Error:      	Not equal: 
        	            	expected: [][]string{[]string{"116", "1", "2", "true"}, []string{"116", "2", "1", "true"}}
        	            	actual  : [][]string{[]string{"116", "1", "2", "false"}, []string{"116", "2", "1", "false"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -5,3 +5,3 @@
        	            	   (string) (len=1) "2",
        	            	-  (string) (len=4) "true"
        	            	+  (string) (len=5) "false"
        	            	  },
        	            	@@ -11,3 +11,3 @@
        	            	   (string) (len=1) "1",
        	            	-  (string) (len=4) "true"
        	            	+  (string) (len=5) "false"
        	            	  }
        	Test:       	TestTenantStatusAPI/index_usage_stats
    --- FAIL: TestTenantStatusAPI/index_usage_stats (89.17s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantStatusAPI.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/ccl/serverccl/statusccl/statusccl_test.TestTenantStatusAPI [failed](https://tanzanite.cluster.engflow.com/invocations/default/6f2cff4d-c2d3-4903-8f72-e74af552ca31?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3NlcnZlcmNjbC9zdGF0dXNjY2w6c3RhdHVzY2NsX3Rlc3Q=) on master @ [c316d6a615fa02a05357a20bf03b8a197ab27810](https://github.com/cockroachdb/cockroach/commits/c316d6a615fa02a05357a20bf03b8a197ab27810):


```
=== RUN   TestTenantStatusAPI
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestTenantStatusAPI3088112816
    tenant_status_test.go:141: -- test log scope end --
CREATE UNIQUE INDEX test_pkey ON test_db1.public.test USING btree (k ASC)test logs left over in: outputs.zip/logTestTenantStatusAPI3088112816
--- FAIL: TestTenantStatusAPI (938.58s)
=== RUN   TestTenantStatusAPI/reset_index_usage_stats
    --- FAIL: TestTenantStatusAPI/reset_index_usage_stats (18.83s)
=== RUN   TestTenantStatusAPI/reset_index_usage_stats/http
    sql_runner.go:87: error executing query="\nCREATE TABLE test (\n  k INT PRIMARY KEY,\n  a INT,\n  b INT,\n  INDEX(a)\n)\n" args=[]: pq: relation "defaultdb.public.test" already exists
        --- FAIL: TestTenantStatusAPI/reset_index_usage_stats/http (2.45s)
=== RUN   TestTenantStatusAPI/reset_index_usage_stats/sql-cli
    tenant_status_test.go:726: query '
        SELECT
          table_id,
          index_id,
          total_reads,
          extract_duration('second', now() - last_read) < 5
        FROM
          crdb_internal.index_usage_statistics
        WHERE
          table_id = 104': expected:
        104, 1, 2, true
        104, 2, 1, true
        
        got:
        104, 1, 2, false
        104, 2, 1, false
        
        --- FAIL: TestTenantStatusAPI/reset_index_usage_stats/sql-cli (16.37s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantStatusAPI.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/serverccl/statusccl/tenant_status_test.go
+++ b/pkg/ccl/serverccl/statusccl/tenant_status_test.go
@@ -45,6 +45,9 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+// This value is arbitrary and needs to be enough in case of slow tests.
+const LastReadThresholdSeconds = 30
+
 func TestTenantStatusAPI(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	s := log.ScopeWithoutShowLogs(t)
@@ -713,17 +716,29 @@ SELECT
   table_id,
   index_id,
   total_reads,
-  extract_duration('second', now() - last_read) < 5
+  extract_duration('second', now() - last_read)
 FROM
   crdb_internal.index_usage_statistics
 WHERE
   table_id = ` + testTableIDStr
 				// Assert index usage data was inserted.
-				expected := [][]string{
-					{testTableIDStr, "1", "2", "true"}, // Primary index
-					{testTableIDStr, "2", "1", "true"},
+				expected := []struct {
+					tableID    string
+					indexID    string
+					totalReads string
+				}{
+					{tableID: testTableIDStr, indexID: "1", totalReads: "2"},
+					{tableID: testTableIDStr, indexID: "2", totalReads: "1"},
 				}
-				cluster.TenantConn(serverccl.RandomServer).CheckQueryResults(t, query, expected)
+				rows := cluster.TenantConn(serverccl.RandomServer).QueryStr(t, query)
+				for idx, e := range expected {
+					require.Equal(t, e.tableID, rows[idx][0])
+					require.Equal(t, e.indexID, rows[idx][1])
+					require.Equal(t, e.totalReads, rows[idx][2])
+					lastReadDurationSec, err := strconv.Atoi(rows[idx][3])
+					require.NoError(t, err)
+					require.LessOrEqualf(t, lastReadDurationSec, LastReadThresholdSeconds, "Last Read was %ss ago, expected less than 30s", lastReadDurationSec)
+				}
 			}
 
 			// Reset index usage stats.
@@ -1047,25 +1062,33 @@ SELECT
   table_id,
   index_id,
   total_reads,
-  extract_duration('second', now() - last_read) < 5
+  extract_duration('second', now() - last_read)
 FROM
   crdb_internal.index_usage_statistics
 WHERE
   table_id = $1
 `
-	actual := testingCluster.TenantConn(2).QueryStr(t, query, testTableID)
-	expected := [][]string{
-		{testTableIDStr, "1", "2", "true"},
-		{testTableIDStr, "2", "1", "true"},
+	expected := []struct {
+		tableID    string
+		indexID    string
+		totalReads string
+	}{
+		{tableID: testTableIDStr, indexID: "1", totalReads: "2"},
+		{tableID: testTableIDStr, indexID: "2", totalReads: "1"},
+	}
+	rows := testingCluster.TenantConn(2).QueryStr(t, query, testTableID)
+	for idx, e := range expected {
+		require.Equal(t, e.tableID, rows[idx][0])
+		require.Equal(t, e.indexID, rows[idx][1])
+		require.Equal(t, e.totalReads, rows[idx][2])
+		lastReadDurationSec, err := strconv.Atoi(rows[idx][3])
+		require.NoError(t, err)
+		require.LessOrEqualf(t, lastReadDurationSec, LastReadThresholdSeconds, "Last Read was %ss ago, expected less than 30s", lastReadDurationSec)
 	}
 
-	require.Equal(t, expected, actual)
-
 	// Ensure tenant data isolation.
-	actual = controlledCluster.TenantConn(0).QueryStr(t, query, testTableID)
-	expected = [][]string{}
-
-	require.Equal(t, expected, actual)
+	actual := controlledCluster.TenantConn(0).QueryStr(t, query, testTableID)
+	require.Equal(t, [][]string{}, actual)
 }
 
 func selectClusterSessionIDs(t *testing.T, conn *sqlutils.SQLRunner) []string {

